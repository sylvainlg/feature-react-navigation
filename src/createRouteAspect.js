import React from 'react';
import * as ReactIs from 'react-is';
import { createAspect, launchApp } from 'feature-u';
import has from 'lodash.has';

import verify from './util/verify';

const logf = launchApp.diag.logf.newLogger(
  '- ***feature-react-navigation*** routeAspect: '
);

export default function createRouteAspect(config = {}) {
  // validate parameters
  const check = verify.prefix('createRouteAspect() parameter violation: ');

  check(
    typeof config === 'object',
    `
    config parameter must be an object

    Allowed parameters are : 
    - name type string
    - Navigator type ReactElement
  `
  );

  const { name = 'route', Navigator } = config;

  check(name, 'name is required');
  check(typeof name === 'string', 'name must be a string');

  // Other fields are tested in genesis function

  return createAspect({
    name,
    genesis,
    validateFeatureContent,
    assembleFeatureContent,
    initialRootAppElm,
    config: {
      Navigator,
    },
  });
}

/**
 * Control aspect config
 */
function genesis() {
  logf('genesis() validating required config.Navigator');

  if (!this.config.Navigator)
    return `the ${this.name} aspect requires Navigator to be provided`;

  if (!ReactIs.isValidElementType(this.config.Navigator)) {
    return `the ${this.name} aspect requires Navigator to be a valid React Element in createRouteAspect object parameter !`;
  }
  return null;
}

/**
 * Validate routes configuration
 *
 * @param {Object} feature Feature configuration
 *
 * @private
 */
function validateFeatureContent(feature) {
  if (has(feature, 'route')) {
    if (
      typeof feature.route !== 'object' ||
      typeof feature.route.routes !== 'object'
    ) {
      return 'If a feature expose some routing, routes must be exposed under feature.route.routes param';
    } else {
      for (const key in feature.route.routes) {
        if (has(feature.route.routes, key)) {
          const el = feature.route.routes[key];
          if (!ReactIs.isValidElementType(el.screen)) {
            return 'Features routes must have a valid React elements under screen property';
          }
        }
      }
    }
  }

  return null; // valid
}

/**
 * Accumulate all routes from our features.
 *
 * @param {Fassets} fassets the Fassets object used in cross-feature-communication.
 *
 * @param {Feature[]} activeFeatures - The set of active (enabled)
 * features that comprise this application.
 *
 * @private
 */
function assembleFeatureContent(fassets, activeFeatures) {
  // accumulate all routes from our features
  const r = {};
  activeFeatures.reduce((accum, feature) => {
    if (!has(feature[this.name], 'routes')) {
      return accum;
    }

    const { routes: routeContent } = feature[this.name];
    if (routeContent) {
      Object.assign(r, routeContent);
    }
    return accum;
  }, []);

  // retain for later use
  this.routes = r;
}

/**
 * Inject our `<AppContainer>` in the `rootAppElm`.
 *
 * We use `initialRootAppElm()` because `<AppContainer>` does NOT
 * support children (by design).
 *
 * @param {Fassets} fassets the Fassets object used in cross-feature-communication.
 *
 * @param {reactElm} curRootAppElm - the current react app element root.
 *
 * @return {reactElm} rootAppElm seeded with our `<AppContainer>`.
 *
 * @private
 */
function initialRootAppElm(fassets, curRootAppElm) {
  // no-op if we have NO routes
  if (
    Object.keys(this.routes).length === 0 &&
    this.routes.constructor === Object
  ) {
    // NOTE: for this condition, the appropriate logf.force() is generated (above)
    return curRootAppElm;
  }

  // insure we don't clober any supplied content
  // ... by design, <AppContainer> doesn't support children
  if (curRootAppElm) {
    throw new Error(
      '***ERROR*** Please register routeAspect (from feature-react-navigation) before other Aspects ' +
        'that inject content in the rootAppElm ... <AppContainer> does NOT support children.'
    );
  }

  // seed the rootAppElm with our StateRouter
  logf(
    `initialRootAppElm() introducing <AppContainer> component into rootAppElm`
  );

  return <Navigator routes={this.routes} />;
}
