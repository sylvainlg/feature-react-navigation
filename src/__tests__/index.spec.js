import createRouteAspect from '../createRouteAspect';
import React from 'react';
import { createSwitchNavigator } from '@react-navigation/core';

function createAppContainer() {
  return class AppContainer extends React.Component {};
}
class Test extends React.Component {}

describe('createRouteAspect', () => {
  it('should return an object', () => {
    const aspect = createRouteAspect();

    expect(typeof aspect).toBe('object');
    expect(aspect).toMatchObject(
      expect.objectContaining({
        assembleAspectResources: undefined,
        assembleFeatureContent: expect.any(Function),
        config: expect.any(Object),
        expandFeatureContent: undefined,
        genesis: expect.any(Function),
        initialRootAppElm: expect.any(Function),
        injectRootAppElm: undefined,
        name: expect.any(String),
        validateFeatureContent: expect.any(Function),
      })
    );
  });

  it('name should be route', () => {
    const aspect = createRouteAspect();
    expect(aspect.name).toBe('route');
  });

  it('config should be an object', () => {
    expect(() => createRouteAspect('custom')).toThrow();
  });

  it('name should be custom', () => {
    const aspect = createRouteAspect({ name: 'custom' });
    expect(aspect.name).toBe('custom');
  });

  it("name shouldn't be null", () => {
    expect(() => createRouteAspect(null)).toThrow();
  });

  it("name shouldn't be a number", () => {
    expect(() => createRouteAspect(123)).toThrow();
  });
});

describe('genesis', () => {
  it('should fail without Navigation config property', () => {
    // Fail => return string; success => return null
    const aspect = createRouteAspect();

    expect(aspect.genesis()).toBe(
      'the route aspect requires Navigator to be provided'
    );
  });

  it('should fail without Navigation config property is a proper React element', () => {
    // Fail => return string; success => return null
    const aspect = createRouteAspect({ Navigator: [] });

    expect(aspect.genesis()).toBe(
      'the route aspect requires Navigator to be a valid React Element in createRouteAspect object parameter !'
    );
  });

  it('should success with configuration', () => {
    // Fail => return string; success => return null
    const aspect = createRouteAspect({
      Navigator: ({ routes }) => null,
    });

    expect(aspect.genesis()).toBe(null);
  });
});

describe('validateFeatureContent', () => {
  const aspect = createRouteAspect();

  it("should be ok when feature don't expose route", () => {
    const feature = {};
    expect(aspect.validateFeatureContent(feature)).toBeNull();
  });

  it("shouldn't be ok when feature expose non object route", () => {
    const feature = { route: 'string' };
    expect(aspect.validateFeatureContent(feature)).toBe(
      'If a feature expose some routing, routes must be exposed under feature.route.routes param'
    );
  });

  it("shouldn't be ok when feature don't expose routes parameter in route object", () => {
    const feature = { route: { routes: 'string' } };
    expect(aspect.validateFeatureContent(feature)).toBe(
      'If a feature expose some routing, routes must be exposed under feature.route.routes param'
    );
  });

  it('should be ok when feature expose routes correctly', () => {
    const feature = { route: { routes: { myroute: { screen: Test } } } };
    expect(aspect.validateFeatureContent(feature)).toBeNull();
  });

  it("shouldn't be ok when feature expose routes with no React Element as value", () => {
    const feature = { route: { routes: { myroute: { screen: null } } } };
    expect(aspect.validateFeatureContent(feature)).toBe(
      'Features routes must have a valid React elements under screen property'
    );
  });
});

describe('assembleFeatureContent', () => {
  it('should succeed when there no features', () => {
    const aspect = createRouteAspect();
    aspect.assembleFeatureContent({}, []);
    expect(aspect.routes).toStrictEqual({});
  });

  it('should succeed when there is no routes in any features', () => {
    const aspect = createRouteAspect();
    aspect.assembleFeatureContent({}, [{}, {}]);
    expect(aspect.routes).toStrictEqual({});
  });

  it('should succeed when there is no routes in any features (route property defined)', () => {
    const aspect = createRouteAspect();
    aspect.assembleFeatureContent({}, [{ route: {} }, {}]);
    expect(aspect.routes).toStrictEqual({});
  });

  it('should succeed when there is no routes in any features (route.routes property defined)', () => {
    const aspect = createRouteAspect();
    aspect.assembleFeatureContent({}, [{ route: { routes: {} } }, {}]);
    expect(aspect.routes).toStrictEqual({});
  });

  it('should assemble routes', () => {
    const aspect = createRouteAspect();
    aspect.assembleFeatureContent({}, [
      { route: { routes: { myroute: { screen: Test } } } },
      {},
    ]);
    expect(aspect.routes).toStrictEqual({ myroute: { screen: Test } });
  });

  it('should assemble routes from multiple features', () => {
    const aspect = createRouteAspect();
    aspect.assembleFeatureContent({}, [
      { route: { routes: { myroute: { screen: Test } } } },
      {
        route: { routes: { index: { screen: Test }, home: { screen: Test } } },
      },
    ]);
    expect(aspect.routes).toStrictEqual({
      myroute: { screen: Test },
      index: { screen: Test },
      home: { screen: Test },
    });
  });
});

describe('initialRootAppElm', () => {
  it('should return the same curRootAppElm if no routes are provided', () => {
    const aspect = createRouteAspect();
    aspect.genesis();
    aspect.assembleFeatureContent({}, []);
    expect(aspect.routes).toStrictEqual({});
    // Object.keys(obj).length === 0 && obj.constructor === Object
    expect(Object.keys(aspect.routes).length).toBe(0);
    expect(aspect.initialRootAppElm({}, 'dummy')).toBe('dummy');
  });

  it('should throw an error when routes and curRootAppElm are defined', () => {
    const aspect = createRouteAspect();
    aspect.genesis();
    aspect.assembleFeatureContent({}, [
      { route: { routes: { myroute: { screen: Test } } } },
    ]);
    expect(() => aspect.initialRootAppElm({}, 'dummy')).toThrow();
  });
});

describe('complex working sample', () => {
  const config = {
    navigator: createSwitchNavigator,
    routes: {
      startup: {
        navigator: createSwitchNavigator,
        navigationOptions: {
          initialRouteName: 'app.START',
        },
        routes: ['app.START'],
      },
      unauth: {
        navigator: createSwitchNavigator,
        featureRoutes: ['app-login', 'app-subscription'],
      },
      auth: {
        navigator: createSwitchNavigator,
        routes: {
          realytricky: {
            navigator: createSwitchNavigator,
            routes: ['home'],
          },
        },
        featureRoutes: ['app-home', 'app-profil', 'app-offers'],
      },
    },
  };

  const features = [
    { route: { routes: { home: { screen: Test } } } },
    { route: { routes: { 'app.START': { screen: Test } } } },
    {
      route: {
        routes: {
          'app-login.LOGIN': { screen: Test },
          'app-login.LOGIN_SUCCESS': { screen: Test },
        },
      },
    },
    {
      route: {
        routes: {
          'app-subscription.START': { screen: Test },
          'app-subscription.FORM': { screen: Test },
        },
      },
    },
    {
      route: {
        routes: {
          'app-home.HOME': { screen: Test },
        },
      },
    },
    {
      route: {
        routes: {
          'app-profil.SETTINGS': { screen: Test },
          'app-profil.SECURITY': { screen: Test },
        },
      },
    },
    {
      route: {
        routes: {
          'app-offers.LIST': { screen: Test },
          'app-offers.DETAILS': { screen: Test },
        },
      },
    },
  ];

  it('should match the snapshot', () => {
    const aspect = createRouteAspect({
      navigationPattern: config,
      createAppFunction: createAppContainer,
    });

    aspect.genesis();
    aspect.assembleFeatureContent({}, features);
    expect(aspect.initialRootAppElm({}, undefined)).toMatchSnapshot();
  });
});
