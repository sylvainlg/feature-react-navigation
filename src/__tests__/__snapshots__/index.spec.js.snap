// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex working sample should match the snapshot 1`] = `
<Navigator
  routes={
    Object {
      "app-home.HOME": Object {
        "screen": [Function],
      },
      "app-login.LOGIN": Object {
        "screen": [Function],
      },
      "app-login.LOGIN_SUCCESS": Object {
        "screen": [Function],
      },
      "app-offers.DETAILS": Object {
        "screen": [Function],
      },
      "app-offers.LIST": Object {
        "screen": [Function],
      },
      "app-profil.SECURITY": Object {
        "screen": [Function],
      },
      "app-profil.SETTINGS": Object {
        "screen": [Function],
      },
      "app-subscription.FORM": Object {
        "screen": [Function],
      },
      "app-subscription.START": Object {
        "screen": [Function],
      },
      "app.START": Object {
        "screen": [Function],
      },
      "home": Object {
        "screen": [Function],
      },
    }
  }
/>
`;
